xlators/features/index/src/index.h

```diff
@@ -55,7 +55,7 @@ typedef struct index_priv {
     dict_t *dirty_watchlist;
     dict_t *pending_watchlist;
     dict_t *complete_watchlist;
-    int64_t pending_count;
+    int32_t pending_count;
```

Comment:
As per my understanding reading a variable in C translates to LOAD instruction which is supposed to load the value from memory to a register. I am not sure if any optimizations can avoid this and reuse an existing register with older value. @xhernandez @mykaul @amarts Can one of you confirm if there is a problem?

Reply:
Reads and writes smaller or equal to the CPU word size and properly aligned in memory are atomic in all architectures I know. This means that even if there are multiple threads accessing and modifying the memory location, it's guaranteed that the read value will be equal to one of the written values (no guarantee about which one), but it will never have a part from one write and a part from another write.

That said, this is not enough for most cases, because a write by one thread may not be seen by a read in another thread instantly and, even worse, a read or write made before or after the atomic access may be seen after or before that access (i.e. in reverse order). If we need to ensure order for reads and writes, memory barriers are necessary. On x86_64, atomic operations (not simple reads and writes) have an implicit memory barrier. But even then it depends on how your code works. Atomic operations cannot always replace a mutex.

For this particular case, I see that `pending_count` is updated conditionally in two steps, so, as it's now, I don't think reads and writes can be done without a lock (a write without a lock in another thread could modify the intermediate value, or a read could get the intermediate value, which is not wanted). However the logic seems quite simple. It's possible that with a better representation of the counter, or if it's possible to be sure that intermediate reads/writes won't happen by program's logic, all management can be done without locks and using atomic operations, but it's not straight forward.

At the very least, you should use CMM_STORE_SHARED()/CMM_LOAD_SHARED() (from urcu lib, or an equivalent) when writing and reading `pending_count` outside a lock to be sure the needed barriers are added. But as I said this might not enough.

Without proper analysis, and given that the operations done inside the lock are extremely fast, the probability of lock contention here should be minimal. Without contention, a mutex lock is equal to a single atomic operation, and a mutex unlock is equal to an atomic write. So I would say this is not too expensive in this case.

I found this [free book](https://cdn.kernel.org/pub/linux/kernel/people/paulmck/perfbook/perfbook.html) some time ago that has some interesting explanations about synchronization and memory ordering. Specially interesting the apendix C to understand how the hardware works (at a high level) and why barriers are needed.

Reply:
All that @xhernandez  wrote, plus:
Looking at this structure, there is room for improvement, What you can see is that the lock object is not in the same cacheline (or close to) the pending_count. In this specific case, if you could put the lock, index and pending_count in the same cacheline, that'd be great - and in the order they are accessed. Currently:
```
struct index_priv {
        char *                     index_basepath;       /*     0     8 */
        char *                     dirty_basepath;       /*     8     8 */
        uuid_t                     index;                /*    16    16 */
        gf_lock_t                  lock;                 /*    32    40 */
        /* --- cacheline 1 boundary (64 bytes) was 8 bytes ago --- */
        unsigned char              internal_vgfid[3][16]; /*    72    48 */
        struct list_head           callstubs;            /*   120    16 */
        /* --- cacheline 2 boundary (128 bytes) was 8 bytes ago --- */
        pthread_mutex_t            mutex;                /*   136    40 */
        pthread_cond_t             cond;                 /*   176    48 */
        /* --- cacheline 3 boundary (192 bytes) was 32 bytes ago --- */
        dict_t *                   dirty_watchlist;      /*   224     8 */
        dict_t *                   pending_watchlist;    /*   232     8 */
        dict_t *                   complete_watchlist;   /*   240     8 */
        int64_t                    pending_count;        /*   248     8 */
        /* --- cacheline 4 boundary (256 bytes) --- */
        pthread_t                  thread;               /*   256     8 */
        _Bool                      down;                 /*   264     1 */

        /* XXX 7 bytes hole, try to pack */

        gf_atomic_int64_t          stub_cnt;             /*   272     8 */
        int32_t                    curr_count;           /*   280     4 */

        /* size: 288, cachelines: 5, members: 16 */
        /* sum members: 277, holes: 1, sum holes: 7 */
        /* padding: 4 */
        /* last cacheline: 32 bytes */
};
```


If we do see contention on the lock, it may be worthwhile to using it to protect only the index and use a different method for the pending_count - but I was not sure atomic operations can be sufficient as you are doing multiples ops on them within a lock - atomic ops don't replace those? (So perhaps a lock for it is still required, but doesn't have to be the same lock as the index?)


(As for the book Xavi recommended - it just reminds me why computer architecture was my favorite, by far, topic in the university. It's fascinating (to me at least) the tricks done to accelerate computing, and what we need to know so we won't be surprised by it). I also recommend https://akkadia.org/drepper/cpumemory.pdf

Reply:
@pranithk  Based on the explanations above, I think it is safe to retain the LOCK for read as well and take this particular fix in. We could address the lockless optimization and structure alignment to prevent cacheline miss in  separate patch(es).